api/v1beta2/awsmachine_types.go

```diff
@@ -64,6 +64,7 @@ const (
 )
 
 // AWSMachineSpec defines the desired state of an Amazon EC2 instance.
+// +kubebuilder:validation:XValidation:rule="!has(self.capacityReservationId) || !has(self.marketType) || self.marketType != 'Spot'",message="capacityReservationId may not be set when marketType is Spot"
```

Comment:
Should this also cover when `SpotMarketOptions` is not nil?

Similarly to what you do below:
```
		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec.awsLaunchTemplate.spotMarketOptions"), "setting spotMarketOptions and CapacityReservationID cannot be used together"))
```